lib/_http_client.js

```diff
@@ -310,28 +334,37 @@ ClientRequest.prototype._implicitHeader = function _implicitHeader() {
                     this[kOutHeaders]);
 };
 
-ClientRequest.prototype.abort = function abort() {
-  if (!this.aborted) {
-    process.nextTick(emitAbortNT.bind(this));
-  }
+ClientRequest.prototype.destroy = destroy;
+ClientRequest.prototype._destroy = function(err, cb) {
   this.aborted = true;
+  process.nextTick(emitAbortNT, this);
 
   // If we're aborting, we don't care about any more response data.
   if (this.res) {
     this.res._dump();
   }
 
-  // In the event that we don't have a socket, we will pop out of
-  // the request queue through handling in onSocket.
-  if (this.socket) {
+  if (this.upgradeOrConnect) {
+    // We're detached from socket.
+    cb(err);
+  } else if (this.socket) {
     // in-progress
-    this.socket.destroy();
+    this.socket.destroy(err, cb);
+  } else if (this[kWritableState].socketPending) {
+    // In the event that we don't have a socket, we will pop out of
+    // the request queue through handling in onSocket.
+    this[kWritableState].destroyCallback = (er) => cb(er || err);
```

Comment:
I think it's better to wait for the `'socket'` event to be emitted instead of adding one more state parameter: https://nodejs.org/api/http.html#http_event_socket.

Reply:
From a user perspective, I think we should be swallowing that error if `destroy()` was called before hand.

Do we test that calling `destroy()` before we have a socket actually calls `socket.destroy()`?

Reply:
This can be further refactored if/once https://github.com/nodejs/node/pull/29656 is merged. But it basically does the same thing.

Reply:
> I think we should be swallowing that error if destroy() was called before hand.

This confuses me. Isn't it a bit inconsistent with the discussion in https://github.com/nodejs/node/pull/29197.

> Do we test that calling destroy() before we have a socket actually calls socket.destroy()?

Not sure, I will check.

Reply:
I have with with @ronag. Swallowing the error would be inconsistent with `destroy()` behavior.

Reply:
How about we just emit(‘error’) instead of passing it to the callback of destroy? That does not feel like the right place to me. Somebody is calling destroy() without an error, and then an error is emitted.

Reply:
> Somebody is calling destroy() without an error, and then an error is emitted.

Isn't that how the flow for destroy is designed at the moment? I kind of agree with you but I lost that discussion in the other PR.

I would recommend consistency here. I think the current state is what best aligns with current design and expectations.

Reply:
> How about we just emit(‘error’) instead of passing it to the callback of destroy?

I'm not sure that will work? We still need to mute calling the callback (i.e.  `'close'`) until we have emitted the error.

Reply:
> Somebody is calling destroy() without an error, and then an error is emitted

Yes that's how `destroy()` currently works if the error occurs before "close".

Reply:
@mcollina: What is your main concern here? The behaviour or the extra property? We could avoid initialising the property in the state constructor which would avoid it in 99% of cases.

Reply:
> Do we test that calling destroy() before we have a socket actually calls socket.destroy()?

Test added

Reply:
The possibility of deadlocks/having a callback not called in the destroy process makes me nervous. The current logic is significantly simpler, even if less correct, and I feel this might bite us in the future.

Reply:
> The possibility of deadlocks/having a callback not called in the destroy process makes me nervous.

I'm not sure I understand.

> The current logic is significantly simpler, even if less correct, and I feel this might bite us in the future.

I strongly disagree with this. In my opinion this will stagnate the project. Shouldn't we strive after correctness and consistency?

We have similar scenarios in a lot of different places and I have pending PR's with efforts on this. If this is the way we go then I'd like to ask #29656 to be reconsidered to achieve some form of consistency.

Reply:
> Shouldn't we strive after correctness and consistency?

No. We should strive to not break current users, and then consistency. Correctness it's extremely hard to measure because there are no specifications for any of this.

Reply:
> We should strive to not break current users

How does this break current users?